Source/dafny.msbuild/DafnyVersionCheckTask.cs

```diff
@@ -0,0 +1,157 @@
+            var enforceCommit = !string.IsNullOrEmpty(DafnyVersionCommit);
+            var enforceTag = !string.IsNullOrEmpty(DafnyVersionTag);
+            if (!(enforceVersion || enforceCommit || enforceTag)) return Success(foundDafnyVersion);
+            if (new[] {enforceVersion, enforceCommit, enforceTag}.Count(b => b) > 1) {
+                return Fail("The DafnyVersion, DafnyVersionCommit, and DafnyVersionTag properties are mutually exclusive");
+            }
+
+            if (enforceVersion) return DafnyVersion == foundDafnyVersion
+                ? Success(foundDafnyVersion)
+                : Fail($"Expected Dafny version \"{DafnyVersion}\", found \"{foundDafnyVersion}\"");
+
+            var expectedRepoHead = enforceCommit ? $"commit {DafnyVersionCommit}" : $"tag {DafnyVersionTag}";
+            if (string.IsNullOrEmpty(DafnyBinariesDir))
+            {
+                return Fail("DafnyBinariesDir must be set to a directory within a Git repo in order to check DafnyVersionCommit or DafnyVersionTag");
+            }
+            if (IsGitWorkingTreeDirty(DafnyBinariesDir))
+            {
+                return Fail($"Expected Dafny executable Git repo at {expectedRepoHead}, but working tree or index has changes");
+            }
+
+            var specifiedCommitHash = enforceCommit ? DafnyVersionCommit : GetCommitHashForGitTag(DafnyBinariesDir, DafnyVersionTag);
+            var repoHeadCommitHash = GetGitRepoHeadCommitHash(DafnyBinariesDir);
+            return repoHeadCommitHash != null && repoHeadCommitHash == specifiedCommitHash
+                ? Success(foundDafnyVersion)
+                : Fail($"Expected Dafny executable Git repo at {expectedRepoHead}, but HEAD is at {repoHeadCommitHash}");
+        }
+
+        /**
+         * Prints a success message and returns true.
+         */
+        private bool Success(string version)
+        {
+            Log.LogMessage(MessageImportance.High, "Found Dafny version {0}", version);
+            return true;
+        }
+
+        /**
+         * Prints the given failure message and returns false.
+         */
+        private bool Fail(string message)
+        {
+            Log.LogError(message);
+            return false;
+        }
+
+        /**
+         * Returns the Dafny version (as printed during execution) and full path; or null, if the process does not print its version in a parseable manner.
+         */
+        private static string GetDafnyExecutableVersion(string dafnyExecutablePath)
+        {
+            using var process = new Process
+            {
+                StartInfo = {FileName = dafnyExecutablePath, CreateNoWindow = true, UseShellExecute = false, RedirectStandardOutput = true}
+            };
+            // Dafny prints its version only if given at least one input file. With an invalid file name, it will print its version, then immediately exit.
+            process.StartInfo.ArgumentList.Add("");
```

Comment:
OR, you could just do `dafny /version` :)

Reply:
Great idea :) Fixed in latest commits.